fml/msg_loop_reconfigurable_task_runner.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#define FML_USED_ON_EMBEDDER
+
+#include "flutter/fml/msg_loop_reconfigurable_task_runner.h"
+#include "flutter/fml/message_loop_impl.h"
+#include "flutter/fml/synchronization/waitable_event.h"
+
+namespace fml {
+
+RefPtr<MsgLoopReconfigurableTaskRunner>
+MsgLoopReconfigurableTaskRunner::CreateFromSingleTaskRunner(
+    const RefPtr<TaskRunner>& task_runner) {
+  if (!task_runner) {
+    return nullptr;
+  }
+  auto loop = task_runner->loop_;
+  return MakeRefCounted<MsgLoopReconfigurableTaskRunner>(loop, loop);
+}
+
+RefPtr<MsgLoopReconfigurableTaskRunner> MsgLoopReconfigurableTaskRunner::Create(
+    const RefPtr<TaskRunner>& task_runner_1,
+    const RefPtr<TaskRunner>& task_runner_2) {
+  if (!task_runner_1 || !task_runner_2) {
+    return nullptr;
+  }
+  auto loop_1 = task_runner_1->loop_;
+  auto loop_2 = task_runner_2->loop_;
+  return MakeRefCounted<MsgLoopReconfigurableTaskRunner>(loop_1, loop_2);
+}
+
+MsgLoopReconfigurableTaskRunner::MsgLoopReconfigurableTaskRunner(
+    const fml::RefPtr<MessageLoopImpl>& loop_1,
+    const fml::RefPtr<MessageLoopImpl>& loop_2)
+    : TaskRunner(nullptr /* loop implemenation*/),
```

Comment:
To address the issue of timed tasks finishing on the correct thread (described below), I suggest moving this switching logic to the message loop impl instead. You can create a subclass that itself takes two MessageLoopImpls and switches the thread assignment on the two when the thread configuration is changed. That way, timed tasks posted on one message loop will get executed on expiry on the new thread. You will be able to do this without latching also.

Reply:
Another advantage of this approach could be that you dont need to change the public API of the shell and its components to vend these new `MsgLoopReconfigurableTaskRunner `s. Instead, you can keep the same task runner interface stable.